Add a wrapper for CancellationTokenSource with a resettable timeout

Brar opened the pull request:
Since resetting a timeout on a CancellationTokenSource is a pretty
common thing to do and that'll probably never be supported out of the
box (https://github.com/dotnet/runtime/issues/4694), put the uglyness of
resetting the timeout in one place so that we don't have to repeat it
all the time.

PS: I take suggestions for the name of the wrapper. I personally don't really care.

YohDeadfall commented on src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+    /// we need to make sure that an existing cancellation token source hasn't been cancelled,
+    /// every time we start it (see https://github.com/dotnet/runtime/issues/4694).
+    /// </remarks>
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
```
First think about this is to make it a struct.

roji replied:
Not sure that's important (these get allocated only per physical connection etc.). IMHO it's OK either as a class or struct.

Brar replied:
I'd say that `CancellationTokenSource` is a class too, so we're allocating anyways.
The additional 8 bytes we allocate for the `Timeout` property will not result in any measurable GC pressure.
I'll leave it as a class.

YohDeadfall replied:
You allocate more than 8 bytes, because of the header prefixing each heap allocated object plus a pointer to the source. It also adds one more object with each allocation which will be analyzed by GC and reallocated.

Take a look at `ManualResetValueTaskSourceCore<TResult>`. It's a structure stored in the object owning it. You did the same helper which has only one owner for its lifetime.

Brar replied:
Ok, I still think that adding all the null checks isn't worthwhile.

Also see https://github.com/npgsql/npgsql/pull/3215#discussion_r499537774

YohDeadfall commented on src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
+    {
+        public TimeSpan Timeout { get; set; }
+        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
```
This would make the field nullable which is a good movement.

Brar replied:
I don't think that this would be a good movement.
It would unnecessarily add more logic to the wrapper.
We'd need a null check on every access to the wrapped `CancellationTokenSource`, even in cases where we know for sure that it can't be null.
We can always make the whole wrapper nullable.

YohDeadfall commented on src/Npgsql/ConnectorPool.Multiplexing.cs:
```diff
@@ -68,8 +68,13 @@ async Task MultiplexingWriteLoop()
             var timeout = _writeCoalescingDelayTicks / 2;
-            var timeoutTokenSource = new CancellationTokenSource();
-            var timeoutToken = timeout == 0 ? CancellationToken.None : timeoutTokenSource.Token;
+            var timeoutToken = CancellationToken.None;
+            TimeoutCancellationTokenSourceWrapper? timeoutTokenSource = null;
+            if (timeout != 0)
+            {
+                timeoutTokenSource = new TimeoutCancellationTokenSourceWrapper(TimeSpan.FromTicks(timeout));
+                timeoutToken = timeoutTokenSource.Token;
+            }
```
Now it's too long, but with suggested changes it will be like:

```csharp
var timeoutTokenSource = new TimeoutCancellationTokenSourceWrapper(TimeSpan.FromTicks(timeout));
var timeoutToken = timeoutTokenSource.Token;
```

roji replied:
Note that this method is active for the lifetime of the program, so there's no need to avoid allocating TimeoutCancellationTokenSourceWrapper if no timeout is defined (this is why the code was as it was). That would make this non-nullable again, which is a bit simpler.

Brar replied:
Yeah, you're right.
I'll revert to the non-nullable variant.
That tiny program-wide allocation doesn't matter.

YohDeadfall commented on src/Npgsql/ConnectorPool.Multiplexing.cs:
```diff
@@ -188,14 +193,9 @@ async Task MultiplexingWriteLoop()
-                        if (timeoutTokenSource.IsCancellationRequested)
+                        if (timeoutTokenSource != null)
                         {
-                            timeoutTokenSource.Dispose();
-                            timeoutTokenSource = new CancellationTokenSource(timeoutTimeSpan);
+                            timeoutTokenSource.Start();
                             timeoutToken = timeoutTokenSource.Token;
```
This call site then will be simplified to:
```csharp
timeoutToken = timeoutTokenSource.Start();
```

Brar replied:
You mean that `Start()` should return the token?
Good idea!

YohDeadfall commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -298,6 +298,9 @@ internal NpgsqlConnector(NpgsqlConnection connection)
             ConnectionString = connectionString;
             PostgresParameters = new Dictionary<string, string>();
             Transaction = new NpgsqlTransaction(this);
+            CommandCts = new TimeoutCancellationTokenSourceWrapper(settings.CancellationTimeout > 0
+                ? TimeSpan.FromSeconds(settings.CancellationTimeout)
+                : Timeout.InfiniteTimeSpan);
```
Check this in the constructor.

roji replied:
I don't think there's any real need to pass the timeout here. We set the timeout every time a command is executed (since it may have a different CommandTimeout), so this isn't really needed.

YohDeadfall commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -1063,7 +1064,9 @@ internal ValueTask<IBackendMessage> ReadMessage(bool async, DataRowLoadingMode d
                     {
                         // TODO: There could be room for optimization here, rather than the async call(s)
-                        ReadBuffer.Timeout = TimeSpan.FromMilliseconds(InternalCommandTimeout);
+                        ReadBuffer.Timeout = InternalCommandTimeout > 0
+                            ? TimeSpan.FromMilliseconds(InternalCommandTimeout)
+                            : Timeout.InfiniteTimeSpan;
```
Count how many times have you repeated these lines. Makes sense to make a decision in the setter (:

roji replied:
I'd just consider keeping 0 as meaning no timeout, just like before.

Brar replied:
I personally hate it when infinity is represented by `0` because semantically a timespan of `0` or `TimeSpan.Zero` means no time at all (immediately; pretty much the opposite of infinite) and there are a few .NET APIs where 0 means just that.
I also find it a bit dangerous if `timeout == System.Threading.Timeout.InfiniteTimeSpan` equals false in case of an infinite timeout.
While we can't change the bad choices ADO.NET made for the timeouts in the connection string variables years ago, we can choose how they are represented in our internal APIs. That's why I changed it (essentially moving the bad choices up to the ultimate public layer where I can't change them), but since 2 people seem not to share my taste, I'll revert that part.

roji replied:
I agree that 0 is problematic and I do actually share your taste. But this isn't just an ADO.NET thing - [Socket.ReceiveTimeout](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.receivetimeout?view=netcore-3.1#System_Net_Sockets_Socket_ReceiveTimeout) works the same way, and IIRC this goes all the way down to Unix system calls...

Brar replied:
`Socket.ReceiveTimeout` is solomonic in that case:

> The default value is 0, which indicates an infinite time-out period. **Specifying -1 also indicates an infinite time-out period.**

YohDeadfall commented on src/Npgsql/NpgsqlWriteBuffer.cs:
```diff
@@ -122,16 +117,9 @@ public async Task Flush(bool async, CancellationToken cancellationToken = defaul
 
             var finalCt = cancellationToken;
-            if (async && Timeout > TimeSpan.Zero)
+            if (async && Timeout > System.Threading.Timeout.InfiniteTimeSpan)
```
`Timeout` condition can be removed too in that case.

Brar replied:
Yeah, I agree that no current use of the `NpgsqlWriteBuffer` class sets the `Timeout` after object initialization but that's not the case for `NpgsqlReadBuffer` and I'd rather keep the symmetry between them.

vonzshik commented on src/Npgsql/NpgsqlReadBuffer.cs:
```diff
@@ -30,33 +30,28 @@ public sealed partial class NpgsqlReadBuffer : IDisposable
-        /// we need to change it when commands are received.
-        /// Also, used as a timeout for async reads.
-        /// </summary>
-        TimeSpan _currentTimeout;
```
This is definitely something I do not like - we're relying on `TimeoutCancellationTokenSourceWrapper` to never change its `Timeout` property by itself. But I'm open to discussion.

Brar replied:
Well, in case of the buffers, since we want symmetry between the async and the sync API regarding timeouts, I don't see a reason why the instance of the stored `TimeoutCancellationTokenSourceWrapper` should have a different timeout than the instance of the stored `Socket`.

vonzshik replied:
They definitely shouldn't have a different timeout between each other. 
My only problem here is that the `Buffer` doesn't own the `TimeoutCancellationTokenSourceWrapper.Timeout` - only the `TimeoutCancellationTokenSourceWrapper` itself. The same can be said about the `Socket`.

But as this is more of an aesthetic problem (and no one else seems to mind), you can leave it as is.

Brar replied:
I see what you mean but in the actual implementation the `TimeoutCancellationTokenSourceWrapper` instance (`_timeoutCts`) is private and we only expose it's `Timeout` property via `NpgsqlReadBuffer.Timeout`, so it's really pure aesthetics.
The only way to make a mess here, is if you're coding inside the `NpgsqlReadBuffer` class.

roji commented on src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+    /// </remarks>
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
+    {
+        public TimeSpan Timeout { get; set; }
```
We may want to do a Debug.Assert that the timeout doesn't get changed between Start/Stop

Brar replied:
I don't know how I'd implement this without the additional complexity of adding an `_isRunning` field which seems a bit much for a simple `Debug.Assert()`.
Also, changing `Timeout` between Start/Stop is neither unintended nor dangerous.
We simply pick up the new value upon the next call to `Start()` and that's it.

The most dangerous thing one can do with this wrapper is to keep a reference to an old `CancellationToken` after calling `Start()` again, but I don't know how I could prevent that.
You can make this mistake with the raw `CancellationTokenSource` too but it would be more obvious in that case.

roji replied:
>  Also, changing Timeout between Start/Stop is neither unintended nor dangerous.

Maybe, I just don't think we'd ever need to change the timeout for the next usage before the current usage is completed (via Stop), and so if that ever happened, it's very likely it would be a bug. Yeah, I was thinking of an `_isRunning` field under `#if DEBUG`. Not critical.

Brar replied:
> I was thinking of an _isRunning field under #if DEBUG.

After all this is definitely not a bad Idea.
I'll add that ASAP.

roji commented on src/Npgsql/NpgsqlReadBuffer.cs:
```diff
@@ -158,16 +153,9 @@ async Task EnsureLong()
                 {
                     var finalCt = cancellationToken;
-                    if (async && Timeout > TimeSpan.Zero)
+                    if (async && Timeout > System.Threading.Timeout.InfiniteTimeSpan)
```
Is this correct? InfiniteTimeSpan is -1 millisecond, so this will enter the condition for 0.

Brar replied:
You're right.

Brar replied:
> You're right.
... kind of...

As it stands we'd convert values of 0 and lower to `Timeout.InfiniteTimeSpan` in `NpgsqlConnector` whenever we set the value for `Timeout` so a value of 0 in the `ConnectionString` would end up as -1 here.

roji commented on src/Npgsql/NpgsqlReadBuffer.cs:
```diff
@@ -107,7 +102,7 @@ internal TimeSpan Timeout
             Underlying = stream;
             _underlyingSocket = socket;
-            _currentTimeout = TimeSpan.Zero;
+            _timeoutCts = new TimeoutCancellationTokenSourceWrapper(System.Threading.Timeout.InfiniteTimeSpan);
```
nit: can do `using InfiniteTimeSpan = System.Threading.Timeout.InfiniteTimeSpan` above.

Brar replied:
Thanks!
Anyhow, that'll no longer be relevant once I revert to previous infinity semantics.

Brar replied:
This actually doesn't work as you can only change type names this way and not property names, but `using static System.Threading.Timeout;` did the trick (and it only imports the `InfiniteTimeSpan` and the `Infinite` property, in case you're afraid of spamming the global namespace).

roji commented on src/Npgsql/ConnectorPool.Multiplexing.cs:
```diff
@@ -68,8 +68,13 @@ async Task MultiplexingWriteLoop()
+            TimeoutCancellationTokenSourceWrapper? timeoutTokenSource = null;
+            if (timeout != 0)
+            {
+                timeoutTokenSource = new TimeoutCancellationTokenSourceWrapper(TimeSpan.FromTicks(timeout));
```
Note that this method is active for the lifetime of the program, so there's no need to avoid allocating TimeoutCancellationTokenSourceWrapper if no timeout is defined (this is why the code was as it was).

roji commented on src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+        /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
+        /// and make sure that it hasn't been cancelled yet
+        /// </summary>
+        public void Start()
```
This should probably just return the CancellationToken (it may be the same, or different if we allocated a new CTS).

roji reviewed (commented):
I'm not clear on the need to change the infinite timeout value from zero to InfiniteTimeSpan (-1 millisecond), it seems to add more checks and conditions more than anything else. Any reason we need to do this?

roji commented:
Thanks for this @Brar, it's a really nice cleanup.

roji commented on src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs:
```diff
@@ -0,0 +1,66 @@
+        /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
+        /// to <see cref="System.Threading.Timeout.InfiniteTimeSpan"/>
+        /// </summary>
+        public void Stop() => _timeoutCts.CancelAfter(System.Threading.Timeout.InfiniteTimeSpan);
```
I think that's supposed to be a pure internal implementation detail, so I can't really see what use this would be outside the wrapper...

vonzshik replied:
I guess you're right and I'm just getting ahead of myself. I'll take another look if/when it would be really needed.

Brar replied:
I don't know how to add this without some overhead so I think we should do that if and when there's an actual use case that can benefit from that.
It's a purely internal API so we can change it whenever we want to.

Brar commented on src/Npgsql/NpgsqlCommand.cs:
```diff
@@ -1156,12 +1158,6 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
 
                         if (cancellationToken.CanBeCanceled)
                         {
-                            if (connector.CommandCts.IsCancellationRequested)
```
Oh, that was indeed an oversight.
I'll add the `IsCancellationRequested` property to the wrapper and use it here.

Brar commented:
> I'm not clear on the need to change the infinite timeout value from zero to InfiniteTimeSpan (-1 millisecond), it seems to add more checks and conditions more than anything else. Any reason we need to do this?

I've given my reasons in https://github.com/npgsql/npgsql/pull/3215#discussion_r499602526 but I agree that it's adding checks so I'll revert that part.

Brar commented:
I think I've addressed all feedback that we agree on and also reverted the part that changed the representation of infinity in the buffers, although - complexity wise - this doesn't bring much to the table.

The unresolved conversations are a bit unclear and generally seem less important to me.

I've left the commits intact so you can follow along the changes.

roji commented on src/Npgsql/NpgsqlCommand.cs:
```diff
@@ -1159,7 +1161,7 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
                             {
                                 connector.CommandCts.Dispose();
-                                connector.CommandCts = new CancellationTokenSource();
+                                connector.CommandCts = new TimeoutCancellationTokenSourceWrapper();
```
This is a bit smelly to me - the whole point of this wrapper is to encapsulate all the checks and reallocation within it. A Restart method could help (the Stopwatch API is a good inspiration source here).

Brar replied:
Yeah, I thought about a method that resets the wrapper to definitely contain an uncancelled `CancellationTokenSource` but without actually starting the timeout.
Like calling `Start()` and then immediately calling `Stop()`.

I'll add that and change the code here.

roji replied:
Yeah. Maybe Reset or Restart (though see discussion below about Restart)

roji commented on src/Npgsql/NpgsqlCommand.cs:
```diff
@@ -832,7 +832,9 @@ void ValidateParameters(ConnectorTypeMapper typeMapper)
         void BeginSend(NpgsqlConnector connector)
         {
-            connector.WriteBuffer.Timeout = TimeSpan.FromSeconds(CommandTimeout);
+            connector.WriteBuffer.Timeout = CommandTimeout > 0
```
Is this still necessary?

Brar replied:
You're right.
That's an oversight when reverting the changed timeout logic.

Brar replied:
Actually our timespans are an absolute mess.
I went ahead and annotated some of the values with documentation comments.

roji commented on src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs:
```diff
@@ -0,0 +1,72 @@
+        /// </summary>
+        public CancellationToken Start()
+        {
+            _timeoutCts.CancelAfter(Timeout);
```
I'd again add a DebugAssert that we're not already running.

Brar replied:
Ok. I'll do that.

roji commented on src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs:
```diff
@@ -0,0 +1,72 @@
+        /// <summary>
+        /// Reset the timeout on the wrapped <see cref="CancellationTokenSource"/>
+        /// </summary>
+        public void Restart() => _timeoutCts.CancelAfter(Timeout);
```
What happens if _timeoutCts expired in the meantime? This looks unsafe...

Brar replied:
The Idea is exactly to restart the timeout *without* changing `_timeoutCts` so that we can detect the fact that it had already timed out in the next iteration.
In other words restart the timer without resetting `_timeoutCts`.

roji replied:
Ah, I see how it's used now, sorry.

I'd at least consider renaming this because it's not at all clear :sweat_smile: Maybe RestartUnlessExpired?

roji replied:
Another (probably better) option is simply to not do this pattern, and simply do ThrowIfCancellationRequested on the token (in ReadBuffer.EnsureLong), and then Stop/Start it (possibly via a method called Restart) - there's no real reason for this pattern of restarting only if it hasn't expired.

Brar replied:
I still see a use case for this method.
It is very slim which is good for tight loops in the hot path and not totally insane if you know what you're doing.
I've named it `RestartTimeoutWithoutReset()` now and extended the documentation comment.

roji replied:
Is there an actual hot path at the moment where we think it's needed?

Brar replied:
Well, we're talking about the [loop where we're reading from the stream in `NpgsqlReadBuffer`](https://github.com/npgsql/npgsql/blob/3c70e78e60511d6b66a01769a75d86caabdc162b/src/Npgsql/NpgsqlReadBuffer.cs#L199) which I'd consider hot path.

Brar commented:
Let's continue the discussion in #3222

Brar commented on src/Npgsql/NpgsqlCommand.cs:
```diff
@@ -1172,8 +1174,7 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
                                 {
                                     cmd.Cancel(true);
-                                    if (cn.Settings.CancellationTimeout > 0)
-                                        cn.CommandCts.CancelAfter(cn.Settings.CancellationTimeout * 1000);
```
Here's the bug where I completely removed the timeout from NpgsqlConnector.CommandCts.
